text/0000-byte-concat.md

```diff
@@ -6,87 +6,83 @@
 # Summary
 [summary]: #summary
 
-Allow the use of `concat!()` to join byte sequences onto an `u8` array,
-beyond the current support for `str` literals.
+Add a macro `concat_bytes!()` to join byte sequences onto an `u8` array,
+the same way `concat!()` currently supports for `str` literals.
 
 # Motivation
 [motivation]: #motivation
 
 `concat!()` is convenient and useful to create compile time `str` literals
-from `str`, `bool`, numeric and `char` literals in the code. This RFC would
-expand this capability to produce `[u8]` instead of `str` when any of its
-arguments is a byte `str` or a byte `char`.
+from `str`, `bool`, numeric and `char` literals in the code. This RFC adds an
+equivalent capability for `[u8]` instead of `str`.
 
 # Guide-level explanation
 [guide-level-explanation]: #guide-level-explanation
 
-Whenever any of the arguments to `concat!()` is a byte literal, its output
-will be a byte literal, and the other arguments will be evaluated on their
-byte contents.
-
-- `str`s and `char`s are evaluated in the same way as `String::as_bytes`,
-- `bool`s are not accepted, use a numeric literal instead,
-- numeric literals passed to `concat!()` must fit in `u8`, any number
-  larger than `std::u8::MAX` causes a compile time error, like the
-  following:  
-```
-error: cannot concatenate a non-`u8` literal in a byte string literal
-  --> $FILE:XX:YY
-   |
-XX |     concat!(256, b"val");
-   |             ^^^ this value is larger than `255`
-```
-- numeric array literals that can be coerced to `[u8]` are accepted, if the
-literals are outside of `u8` range, it will cause a compile time error:
-```
-error: cannot concatenate a non-`u8` literal in a byte string literal
-  --> $FILE:XX:YY
-   |
-XX |     concat!([300, 1, 2, 256], b"val");
-   |              ^^^        ^^^ this value is larger than `255`
-   |              |
-   |              this value is larger than `255`
-```
-
-For example, `concat!(42, b"va", b'l', [1, 2])` evaluates to
+The `concat_bytes!()` macro concatenates literals into a static byte slice.
```

Comment:
Why a slice and not an array?

Reply:
That would make storing the result in a `const` or `static` rather awkward, at least without https://github.com/rust-lang/rfcs/pull/2545. You would have to write

```rust
const FOO: [u8; 5] = concat_bytes!(b"ab", b"123");
```

or in a macro (not super sure whether this actually works):

```rust
const FOO: [u8; concat_bytes!($($args)+).len()] = concat_bytes!($($args)+);
```

Reply:
I think returning an array should be considered as an alternative though; with RFC2545 it would become ergonomic and I'm sure *somebody* has a use case for this 😄

Reply:
b"..." is also a &[u8; N], so concat_bytes!() should imo have the same type. That will auto coerce to a slice.

Reply:
Actually, I just checked and both of

```rust
const FOO: &[u8; 4] = b"true";
const BAR &[u8] = b"true"; // implicit coercion AFAICT
```

work, so I'll update the PR to talk about the output being a byte string literal. IIRC `concat!` also produces a literal rather than an "opaque" (to macros) `&'static str` so this would be consistent.

Reply:
I think what got me confused is the docs for `concat!` talking about a `&'static str` value rather than a literal as the output.